### DISCUSSION TITLE
initialize dual variables for ipopt

### DISCUSSION THREAD

**Discussion by LoreMoretti** [casadi/casadi#3279] [Q&A]
Hello,

I have formulated a MPC problem leveraging the Opti stack and using ipopt as a solver.

Moreover, I have wrapped it into a CasADi function leveraging the opti.to_function() utility.

Is there a way to warm start the solver by providing also the initial guess for dual variables (for example of the previous computation)? 

Ideally, I am wondering if there is a way to pass the initial guess for the dual variables to the CasADi function which wraps the optimization problem.

Thanks a lot for your help,

Lorenzo

**Comment by itbellix** [comment]
Hi @LoreMoretti,
I am not an expert but I believe you might be looking for something like [this](https://casadi.sourceforge.net/users_guide/html/node9.html#SECTION00920000000000000000) (section 9.2.0.0.4, from the CasADi documents)

Best regards,
Italo

**Reply by LoreMoretti** [reply]
Hi @itbellix, Thanks! I will try it out!

**Comment by jgillis** [comment]
`opti.lam_g` can be used both in the input and output argument lists of to_function. `opti.dual()` probably won't work because it is an expression instead of a symbol.

**Reply by LoreMoretti** [reply]
Hi @jgillis, let me try it out! Thanks for your reply.

**Comment by itbellix** [comment]
Hi @LoreMoretti, could you find a solution for how to do this in the end? Thanks!

EDIT: in my case, I am first building my optimization problem with `Opti()`, where my optimization variables are the state trajectory `Xs` and controls `Us` (as `casady.MX.sym()` variables). I have an initial state that is just a parameters, so that, when I generate the CasADi function that encapsulate my optimization problem I do:

```python
p = opti.parameter(2)
F_1 = opti.to_function('F_1', [p], [Xs, Us], ['initial_state'], ['x_opt', 'u_opt'])

x_opt1, u_opt1 = F_1(np.array([0,0]) # correct optimal trajectory given the initial state
```

I am trying to add the initial guesses both for my primal and dual variables like:
```python
F_2 = opti.to_function('F_2', [p, Xs, Us, opti.lam_g], [Xs, Us, opti.lam_g], ['initial_state', 'guess_state_traj', 'guess_controls', 'guess_dual_vars'], ['x_opt', 'u_opt', 'dual_vars'])

x_opt2, u_opt2, _ = F_2(x_0, x_opt1, u_opt1, lam_g0) # where lam_g0 = sol.value(opti.lam_g) after solving the Opti instance once
```

However, I do not observe significant speed up in the solution, and I am unable to check the number of iterations required for convergence from the CasADi function (to check if I am providing the initial guesses correctly).

So, I am not sure if my approach is actually working.

**Reply by S-Dafarra** [reply]
Hi @itbellix! Maybe, also this discussion might have some hints: https://github.com/casadi/casadi/discussions/3411

In general, if you are using ``ipopt`` there are a bunch of parameters you could try touching to see if something improves: https://coin-or.github.io/Ipopt/OPTIONS.html#OPT_Warm_Start

**Reply by itbellix** [reply]
Hi @S-Dafarra, thank you very for jumping in and getting back to this!

Adding the `ipopt` options as reported in https://github.com/casadi/casadi/wiki/FAQ%3A-Warmstarting-with-IPOPT was indeed what I was missing (even if I had found the issue you pointed me at, but apparently not implemented those suggestions...)

Thank you again, now the solution of the problem is quicker as expected if I warm-start it (and, btw, setting the options `'ipopt.print_level': 3` and `'print_time': 3` allows still to print to monitor the iterations, so I could check that they decrease with warm-starting too)!

**Reply by LoreMoretti** [reply]
Sorry, came late to the party.

Thanks @S-Dafarra for pointing to the link.

**Comment by itbellix** [comment]
Just a final comment for other people who are interested in the same application: in #3434 you can find some code that can be useful to run to get the gist of how to warm start the CasADi functions that can be obtained out of `Opti()`

**Comment by diegoferigo** [comment]
Another piece of related documentation is the following: https://www.gams.com/latest/docs/S_IPOPT.html#IPOPT_WARMSTART
